voucher/views.py

```diff
@@ -1,3 +1,195 @@
+        """
+        voucher = get_current_voucher(self.request.user)
+        product_id, coupon_id = json.loads(request.POST["coupon_version"])
+        voucher.coupon_id = coupon_id
+        voucher.save()
+        enroll_url = f"/checkout?product={product_id}&code={voucher.coupon.coupon_code}"
+        return redirect(enroll_url)
+
+
+@login_required
+def resubmit(request):
+    """
+    Prompt user to email voucher after failed voucher parsing
+    """
+    return render(request, "resubmit.html", context=get_js_settings_context(request))
+
+
+@login_required
+def redeemed(request):
+    """
+    Coupon has already been redeemed
+    """
+    return render(request, "redeemed.html", context=get_js_settings_context(request))
+
+
+def get_current_voucher(user):
+    """
+    Get the active voucher for a user
+    Args:
+        user (User): user to return voucher for
+
+    Returns:
+        Voucher: active voucher for user
+    """
+    return user.vouchers.order_by("uploaded").last()
+
+
+def get_eligible_coupon_choices(voucher):
+    """
+    Find exact or partial matching course runs and get valid coupons for them
+
+    Args:
+        voucher (Voucher): a voucher to find courses for
+
+    Returns:
+        list of tuple:
+            list of ('[<product_id>, <coupon_id>]', <course title>) for an eligible coupon / CourseRun match
+    """
+    course_matches = CourseRun.objects.filter(
+        course__readable_id__exact=voucher.course_id_input,
+        course__title__iexact=voucher.course_title_input,
+        start_date__date=voucher.course_start_date_input,
+    )
+    if course_matches.count() > 1:
+        log.error("Found multiple exact CourseRun matches for voucher %s", voucher.id)
+        return []
+    elif not course_matches:
+        # Try partial matching
+        course_matches = CourseRun.objects.filter(
+            Q(course__readable_id__exact=voucher.course_id_input)
+            | Q(course__title__iexact=voucher.course_title_input)
+            | Q(start_date__date=voucher.course_start_date_input)
+        )
+    if not course_matches:
+        # No partial matches found
+        return []
+
+    # Check for valid coupon options and return choices
+    valid_coupons = [
+        get_valid_coupon_versions(
+            match.product.first(),
+            voucher.user,
+            full_discount=True,
+            company=Company.objects.get(id=settings.VOUCHER_COMPANY_ID),
+        )
+        .filter(coupon__voucher=None)
+        .first()
+        for match in course_matches
+    ]
+    eligible_choices = [
+        (
+            json.dumps(
+                (course_matches[i].product.first().id, valid_coupons[i].coupon.id)
+            ),
+            course_matches[i].course.title,
+        )
+        for i in range(len(course_matches))
+        if valid_coupons[i] is not None and course_matches[i].product is not None
+    ]
+    if course_matches and not eligible_choices:
```

Comment:
Still concerned about this approach.  Two users entering a voucher for the same course at about the same time are going to get the same coupon code.  Maybe this could work: in `EnrollView.post`, before saving, check to see if the selected coupon code is already associated with another voucher.  If that's the case, then run `get_valid_coupon_versions` again, to choose a new code to use with the voucher.  Maybe you could add another optional kwarg for `get_valid_coupon_versions` -`voucher_filter=False` - to exclude any coupons associated with vouchers if set to `True`.

Reply:
Yeah we basically need to put a lock on the coupon in between the get and the post. If the user walks away from their computer that could cause problems. The other approach would be to literally put a `locked` or `offered` field on the coupon and set it to true when on the get and then free it up or attach it to voucher on the post. Maybe that's more overhead then necessary. I'll just put the redundant check in post.